### PR TITLE
Make json and yaml available in templates

### LINKED ISSUE #27079: Option to deserialize JSON from last log line in BashOperator and DockerOperator before sending to XCom
### Description

In order to create an XCom value with a BashOperator or a DockerOperator, we can use the option `do_xcom_push` that pushes to XCom the last line of the command logs.

It would be interesting to provide an option `xcom_json` to deserialize this last log line in case it's a JSON string, before sending it as XCom. This would allow to access its attributes later in other tasks with the `xcom_pull()` method. 

### Use case/motivation

See my StackOverflow post : https://stackoverflow.com/questions/74083466/how-to-deserialize-xcom-strings-in-airflow

Consider a DAG containing two tasks: `DAG: Task A >> Task B` (BashOperators or DockerOperators). They need to communicate through XComs.

- `Task A` outputs the informations through a one-line json in stdout, which can then be retrieve in the logs of `Task A`, and so in its *return_value* XCom key if `xcom_push=True`. For instance : `{"key1":1,"key2":3}`

- `Task B` only needs the `key2` information from `Task A`, so we need to deserialize the *return_value* XCom of `Task A` to extract only this value and pass it directly to `Task B`, using the jinja template `{{xcom_pull('task_a')['key2']}}`. Using it as this results in `jinja2.exceptions.UndefinedError: 'str object' has no attribute 'key2'` because *return_value* is just a string.

For example we can deserialize Airflow Variables in jinja templates (ex: `{{ var.json.my_var.path }}`). Globally I would like to do the same thing with XComs.

**Current workaround**: 

We can create a custom Operator (inherited from BashOperator or DockerOperator) and augment the `execute` method:

1. execute the original `execute` method
2. intercepts the last log line of the task
3. tries to `json.loads()` it in a Python dictionnary 
4. finally return the output (which is now a dictionnary, not a string)

The previous jinja template `{{ xcom_pull('task_a')['key2'] }}` is now working in `task B`, since the XCom value is now a Python dictionnary.

```python
class BashOperatorExtended(BashOperator):
    def execute(self, context):
        output = BashOperator.execute(self, context)
        try: 
            output = json.loads(output)
        except:
            pass
        return output

class DockerOperatorExtended(DockerOperator):
    def execute(self, context):
        output = DockerOperator.execute(self, context)
        try: 
            output = json.loads(output)
        except:
            pass
        return output
```

But creating a new operator just for that purpose is not really satisfying..

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by RachitSharma2001**
closes: #27079

It was suggested that a macro be added that allows for the deserialization of json or yaml strings (since yaml is a superset of jsons, I just deserialized using yaml).

I also added tests for this macro in `tests/macro`.

I would also be open to adding this functionality to the system tests if that is needed.

### CODE REVIEW

File: airflow/macros/__init__.py [hunk 1]
```diff
@@ -24,6 +24,7 @@
 from typing import Any
 
 import dateutil  # noqa
+import yaml
```
Comment (Taragolis) [inline airflow/macros/__init__.py]:
```suggestion
import airflow.utils.yaml as yaml
```

File: airflow/macros/__init__.py [hunk 2]
```diff
@@ -85,3 +86,12 @@ def datetime_diff_for_humans(dt: Any, since: DateTime | None = None) -> str:
     import pendulum
 
     return pendulum.instance(dt).diff_for_humans(since)
+
+
+def deserialize(data: str) -> dict[Any, Any]:
```
Comment (Taragolis) [inline airflow/macros/__init__.py]: `deserialize` it is a bit broad name which could mean everything and might confuse users.

File: tests/macros/test_macros.py [hunk 3]
```diff
@@ -89,3 +89,19 @@ def test_ds_format(ds, input_format, output_format, expected):
 def test_datetime_diff_for_humans(dt, since, expected):
     result = macros.datetime_diff_for_humans(dt, since)
     assert result == expected
+
+
+@pytest.mark.parametrize(
+    "ds, expected",
```
Comment (Taragolis) [inline tests/macros/test_macros.py]: `ds` refers to date and known **ds** macro which might confuse devs. `value` or `input_value` more suitable here

**Review by Taragolis — DISMISSED** [review]
Some nits

Also we should add info into the [Templates reference](https://airflow.apache.org/docs/apache-airflow/stable/templates-ref.html#macros) documentation. Otherwise how users would know how to use this macro?

**Comment by RachitSharma2001** [comment]
Sounds good! I have integrated the requested changes in my latest push. 

However, in terms of the documentation, I think it will already automatically be added to the templates reference. This is because this macro is a macro function, specified in [airflow/macros](https://github.com/RachitSharma2001/airflow/blob/airflow_json_loads/airflow/macros/__init__.py) along with `ds_add`, `ds_format`, `datetime_diff_for_humans`). In the bottom of [templates_ref.rst](https://github.com/RachitSharma2001/airflow/blob/airflow_json_loads/docs/apache-airflow/templates-ref.rst), it loads the documentation provided for each function in the that macro file. So I think the documentation I added to the `json_loads` macro (specified in the [comment](https://github.com/RachitSharma2001/airflow/blob/airflow_json_loads/airflow/macros/__init__.py#L92) provided under the function) will actually be linked in the templates reference. Let me know if this is correct or if my understanding is wrong.

**Comment by uranusjr** [comment]
I’d prefer adding two separate macros, one for only JSON and another for YAML support.

**Comment by RachitSharma2001** [comment]
Sounds good! I have created those two separate macros and have rebased. Let me know if any other changes are needed.

**Review by uranusjr — APPROVED** [review]

**Comment by uranusjr** [comment]
Hmm, looking at other things in the module, I wonder if we should simply import `json` and `yaml` into it, so users can just write `{{ json.loads(...) }}` instead of needing to remember a slightly different syntax.

**Comment by potiuk** [comment]
> Hmm, looking at other things in the module, I wonder if we should simply import `json` and `yaml` into it, so users can just write `{{ json.loads(...) }}` instead of needing to remember a slightly different syntax.

Yeah. It woudl be useful to add json/yaml generic imports for many other use cases. The only potential problem is safety of those, but I think those two shoudl be rather safe to add.

**Comment by RachitSharma2001** [comment]
That makes sense. Would it still make sense to keep the tests on json and yaml, or should those be removed?

**Review by uranusjr — APPROVED** [review]
As a follow up, those `noqa` comments should be removed and replaced by an explicit whitelist.

**Comment by RachitSharma2001** [comment]
Hi everyone, it seems like this PR is approved. Is there anything else that should be updated?

**Comment by RachitSharma2001** [comment]
Hi everyone, it seems like this PR has been approved and passes all the tests. I was wondering if there is anything else that should be updated or changed? If so, let me know.

**Comment by RachitSharma2001** [comment]
Hi everyone, I was wondering, since the changes on this branch have been approved, if the branch could be merged into main? Or if there is anything else that needs to be updated or changed, let me know.

**Review by potiuk — APPROVED** [review]

**Pull request merged**